### PR TITLE
GpxSegmentsApproximation C++ port

### PR DESCRIPTION
**Pull request by RZR-UA**

### CODE REVIEW

**Comment by RZR-UA** [comment]
https://github.com/osmandapp/OsmAnd-Issues/issues/2411

File: native/src/gpxSegmentsApproximation.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    double dist = 0;
+    for (int i = start; i < end; i++) {
+        dist += squareRootDist31(
+            pnt->getRoad()->getPoint31XTile(i), pnt->getRoad()->getPoint31YTile(i),
+            pnt->getRoad()->getPoint31XTile(i + 1), pnt->getRoad()->getPoint31YTile(i + 1));
+    }
+    return dist * dist;
+}
+
+const SHARED_PTR<GpxPoint>& GpxSegmentsApproximation::findNextRoutablePoint(RoutePlannerFrontEnd* frontEnd,
```
Comment (ivanPyrohivskyi) [inline native/src/gpxSegmentsApproximation.cpp:126]: SHARED_PTR<GpxPoint>& - just return shared

File: native/src/gpxSegmentsApproximation.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                                                                      double distThreshold,
+                                                                      std::vector<SHARED_PTR<GpxPoint>>& gpxPoints,
+                                                                      int searchStart) {
+    for (int i = searchStart; i < gpxPoints.size(); i++) {
+        if (frontEnd->initRoutingPoint(gpxPoints.at(i), gctx, distThreshold)) {
+            return gpxPoints.at(i);
+        }
+    }
+
+    static SHARED_PTR<GpxPoint> null = nullptr;
```
Comment (ivanPyrohivskyi) [inline native/src/gpxSegmentsApproximation.cpp:137]: !

File: native/src/gpxSegmentsApproximation.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                                     std::vector<SHARED_PTR<GpxPoint>>& gpxPoints);
+
+private:
+    static double minDistResult(SHARED_PTR<RouteSegmentResult>& res, double minDistSqr,
+                                const SHARED_PTR<RouteSegmentPoint>& pnt,
+                                const SHARED_PTR<GpxPoint>& loc);
+
+    static double sumPntDistanceSqr(const SHARED_PTR<RouteSegmentPoint>& pnt, int start, int end);
+
+    static const SHARED_PTR<GpxPoint>& findNextRoutablePoint(RoutePlannerFrontEnd* frontEnd,
```
Comment (ivanPyrohivskyi) [inline native/src/gpxSegmentsApproximation.h:21]: SHARED_PTR& - just return shared

**Comment by ivanPyrohivskyi** [comment]
iOS must be updated
https://github.com/osmandapp/OsmAnd-core/blob/b4873278d0abc2fe8b53adaac91d0a35acae05ad/CMakeLists.txt#L260

**Review by ivanPyrohivskyi — APPROVED** [review]

**Comment by RZR-UA** [comment]
moved to 4.7

**Pull request closed without merging**